python/cudf/cudf/dataframe/numerical.py

```diff
@@ -246,6 +246,33 @@ def round(self, decimals=0):
             data=Buffer(rounded), mask=mask, dtype=self.dtype
         )
 
+    def clip(self, lower=None, upper=None, inplace=False):
+        mask = None
+        if self.has_null_mask:
+            mask = self.nullmask if inplace else self.nullmask.copy()
+
+        if lower is not None and upper is not None:
+            # check value of lower and upper, and switch them if needed
+            if np.isscalar(lower) and np.isscalar(upper):
+                lower, upper = min(lower, upper), max(lower, upper)
+
+        elif lower is None and upper is None:
+            # if both lower and upper are None, return itself or simply copy it
+            if inplace:
+                return self
+            else:
+                return self.copy()
+        elif lower is None:
+            lower = np.NINF
+        else:  # upper is None
+            upper = np.inf
```

Comment:
Using negative and positive infinity won't work for datetime types as Numpy / Numba don't have defined behaviors for this. If self.dtype is a datetime type you'd need to calculate different upper and lowers here.

Reply:
Additionally, we should implement this for datetime columns as well.

Reply:
For `datetime64` types we should use `max(int64)` and `min(int64) + 1` (to avoid NaT values)

Reply:
Got it!

Reply:
I'd better add a clip() function to DatetimeColumn class, right?

Reply:
And maybe also add clip() to CategoricalColumn class and StringColumn class?

Reply:
CategoricalColumn should not be supported, added TypeError.
I met a stubborn numba cuda bug while implementing DatetimeColumn, and StringColumn uses Nvstrings which a little bit complex to deal with. So I raised NotImplementedError for StringColumn and DatetimeColumn